src/pyfreeleh/row/gsheet.py

```diff
@@ -1,412 +1,458 @@
-            self._sheet_name,
-            self._query.build_select(self._selected_columns),
-        )
+    def execute(self) -> List[T]:
+        """Execute the select statement.
+
+        Returns:
+            list: list of rows that matched the given condition.
+        """
+        mapped_columns = []
+        for col in self._selected_columns:
+            mapped_columns.append(self._mapper.to_column(col))
+
+        if "A" not in mapped_columns:
+            mapped_columns.insert(0, "A")
+
+        rows = self._sheet_session.query(self._query.build_select(mapped_columns))
 
         results = []
         for row in rows:
-            row_result = {}
-            for k, v in row.items():
-                row_result[self._field_by_col[k]] = v
-            results.append(row_result)
+            results.append(self._serializer.deserialize(row))
 
         return results
 
 
-class InsertStmt:
+class InsertStmt(Generic[T]):
     def __init__(
         self,
-        wrapper: GoogleSheetWrapper,
-        spreadsheet_id: str,
-        sheet_name: str,
-        columns: List[str],
-        rows: List[Dict[str, str]],
+        sheet_session: GoogleSheetSession,
+        serializer: Serializer[T],
+        rows: List[T],
     ):
-        self._wrapper = wrapper
-        self._spreadsheet_id = spreadsheet_id
-        self._sheet_name = sheet_name
-        self._columns = columns
+        self._sheet_session = sheet_session
+        self._serializer = serializer
         self._rows = rows
 
-        a1_col_mapping = get_a1_column_mapping(columns)
-        self._query = QueryBuilder(a1_col_mapping)
-        self._col_mapping = a1_col_mapping
-        self._field_by_col = {col: field for field, col in self._col_mapping.items()}
+        self._query = GoogleSheetQueryBuilder()
 
     def execute(self) -> None:
+        """Execute the insert statement.
+
+        After a successful insert, all of the `rid` field of the passed in `rows` will be updated.
+
+        Examples:
+            Insert a row.
+
+            >> row = Row(name="cat")
+            >> store.insert([row]).execute()
+            >> row.rid
+            2
+        """
         raw_values = self._get_raw_values()
-        self._wrapper.overwrite_rows(self._spreadsheet_id, A1Range.from_notation(self._sheet_name), raw_values)
+        result = self._sheet_session.overwrite_rows(A1Range.from_notation(self._sheet_session.sheet_name), raw_values)
+
+        # TODO(fata.nugraha): think about how to set rid back.
+        # we should not assume rid is the primary key.
+        for idx, row in enumerate(result.inserted_values):
+            self._rows[idx].rid = int(row[0])
 
     def _get_raw_values(self) -> List[List[str]]:
         raw_values = []
+
         for row in self._rows:
-            raw_values.append(self._to_values_row(row))
-        return raw_values
+            serialized = self._serializer.serialize(row)
+            # TODO(fata.nugraha): assumption A is the pk
```

Comment:
Maybe the seralised result should not contain the `_rid` column first since it's not part of the model definition?

Reply:
The serializer's role here is to convert GViz's output to an instance object.

> `_rid` column first since it's not part of the model definition

Actually it's part of the model definition.